### PR TITLE
Fix host PID detection when CUDA_VISIBLE_DEVICES doesn't start at device 0

### LINKED ISSUE #225: set_task_pid: deterministic "host pid is error" when CUDA_VISIBLE_DEVICES does not start with the pod's device 0 (probe context and poll target mismatch), leading to spurious Device 0 OOM
## Summary

In a container with multiple vGPUs, `set_task_pid()` fails deterministically ("host pid is error!") for any process whose `CUDA_VISIBLE_DEVICES` does not **start with** the container's device 0. The failure is not a race: it reproduces 100% with a single process and no concurrency. After the failure, the process's memory usage is merged into the wrong accounting slot, so allocations that are legal per device get summed against one device's limit and trip a spurious `Device 0 OOM`.

This looks like the root cause behind reports such as Project-HAMi/HAMi#1455 (same log signature during multi-GPU training, closed as stale without analysis).

## Environment

- HAMi v2.7.1 (bundled libvgpu), also traced against HAMi-core `main` sources
- Kubernetes + hami-scheduler, pod with `nvidia.com/gpu: 2`, `nvidia.com/gpumem: 5800` (per device)
- NVIDIA driver 580.82.07, RTX 2080 Ti x2 (originally observed on A100-80GB x8, driver 580.105, verl FSDP world=8)
- PyTorch 2.x (`vllm/vllm-openai:v0.10.2` image), CUDA 12.x

## Reproduction

Inside a single pod that was allocated 2 vGPUs, run one Python process per case (no concurrency at all), each allocating 4096 MiB in 256 MiB steps (well under the 5800 MiB per-device cap):

```bash
CUDA_VISIBLE_DEVICES=<case> python3 -c '
import sys, torch
# optional: torch.cuda.set_device(1)
torch.cuda.init()
bufs = [torch.empty(256 << 20, dtype=torch.uint8, device="cuda") for _ in range(16)]
'
```

Results (3/3 identical across repetitions):

| case | CUDA_VISIBLE_DEVICES | context created on | result |
|---|---|---|---|
| A | `0` | pod dev 0 | ok |
| B | `1` | pod dev 1 | **host pid is error** |
| C | `1,0` | pod dev 1 (default) | **host pid is error** |
| D | `1,0` + `set_device(1)` | pod dev 0 | **host pid is error** |
| E | `0,1` + `set_device(1)` | pod dev 1 | ok |
| F | `0,1` | pod dev 0 | ok |

So the failure condition is exactly: **`CUDA_VISIBLE_DEVICES[0]` != the pod's first device**. Where the user code later places its context is irrelevant (C vs D vs E), because the probe happens during library initialization.

With two workers (CVD=`0` and CVD=`1`, e.g. Ray actors in one pod, or any launcher that narrows CVD per rank), the CVD=`1` worker fails host-pid detection every time, and its allocations are accounted together with the other worker's, producing `Device 0 OOM <sum>/<cap>` while each device is individually within its cap. Observed signatures: `Device 0 OOM 6337593344 / 6081740800` (2080 Ti), `88.79GB / 80GB` (A100 x8).

## Root cause (from source)

In `src/utils.c`, `set_task_pid()`:

1. The before/after process snapshots are taken from the **first CVD-visible device in NVML (physical) enumeration order** — the loop does an unconditional `break` after the first device where `nvml_to_cuda_map(i) >= 0`. For CVD=`1,0` or `0,1` that is the pod's physical device 0.
2. The probe context, however, is created with `cuDevicePrimaryCtxRetain(&pctx, 0)` — CUDA device 0, i.e. **whatever CVD lists first**.

When CVD=`1,0`, the poll watches physical device 0 while the probe context appears on physical device 1, so `getextrapid()` never sees a new PID and returns 0 -> `LOG_ERROR("host pid is error!")`.

When CVD=`1`, physical device 0 is not visible at all, `nvml_to_cuda_map(0)` returns -1 for every polled index that could match, no snapshot is taken (`previous == running == 0`), and `getextrapid()` again returns 0. Same error, different path.

All six cases above, plus the original multi-worker OOM, are consistent with this model.

## Suggested fix

Create the probe context on the same device that is being polled, e.g. retain the primary context for the `cudaDev` found by the map inside the loop instead of hardcoding device 0:

```c
cudaDev = nvml_to_cuda_map(i);
if (cudaDev < 0) continue;
/* poll nvml index i ... */
CHECK_CU_RESULT(cuDevicePrimaryCtxRetain(&pctx, cudaDev));
```

(or equivalently, poll the NVML device that corresponds to CUDA device 0). Either direction removes the mismatch and also fixes the CVD=`1` path, since the polled device and the probe device are then guaranteed to be the same.

## Workarounds we validated

- Keep the pod's first device first in CVD (e.g. `0,1` is safe, `1,0` is not), or
- unset CUDA_VISIBLE_DEVICES in the container and select devices with `torch.cuda.set_device()` per process, or
- use a 1-GPU-per-pod layout so every process's CVD starts with that pod's device 0.

Happy to provide the full reproduction manifests (Kubernetes pod specs) if useful.

### PR DESCRIPTION
**Pull request by ihnokim**
Fixes #225 (closed by the lifecycle policy while the bug is still present on `main`).

## Changes?

- `set_task_pid()` now pairs the probe context and the polled NVML device by UUID: the probe context is created on CUDA device 0, and the NVML device whose UUID matches (`cuDeviceGetUuid` vs `nvmlDeviceGetUUID`) is used for both process snapshots.
- If no NVML device matches the UUID, it logs a warning and falls back to polling NVML device 0 (the previous behavior).
- Error paths now release the retained probe context; previously it leaked for the lifetime of the process whenever detection failed (the leak observed during the #230 testing).
- Not in scope: the `nvml_preInit()` map reset itself (see below). This PR only makes `set_task_pid()` independent of the map; other consumers of the map are unchanged.

## Why We Need?

- Whenever `CUDA_VISIBLE_DEVICES` does not start with the container's device 0, host-PID detection fails deterministically (`host pid is error!`), per-process accounting is merged into the wrong slot, and legal allocations trip spurious per-device OOMs. This is the failure analyzed in #225.
- Direct cause: the probe context is created with `cuDevicePrimaryCtxRetain(&pctx, 0)`, which is CUDA device 0, while the before/after NVML snapshots are taken on the first NVML device that has a valid entry in `cuda_to_nvml_map`. With such a `CUDA_VISIBLE_DEVICES`, those are different physical GPUs, so the probe PID never appears in the snapshot diff.
- Root cause of why the map cannot be used here: `nvml_preInit()` resets `cuda_to_nvml_map_array` to identity, and it is triggered via `pthread_once` by the `nvmlInit()` call at the top of `set_task_pid()` itself. By the time the polling loop consults the map, it is an identity map regardless of what `parse_cuda_visible_env()` computed earlier in `postInit()`.
- This is also why the approach in #230 (keep the probe on the `cudaDev` found through the map) did not resolve the failure when we tested it: with an identity map that `cudaDev` is always 0, which is behaviorally identical to the hardcoded value. It is consistent with the `current processes num = 0 0` runs reported in the #230 thread with that patch applied. The unresolved `SET_TASK_PID FAILED` case on driver 580.65.06 from that thread matches this root cause as well, but we could not verify on that exact driver.
- The UUID pairing holds regardless of the state of the map or of the initialization order.

## Tests?

Environment: Kubernetes + HAMi v2.7.1 (hami-scheduler), pod with `nvidia.com/gpu: 2`, `nvidia.com/gpumem: 5800`; node: 2x NVIDIA A100-PCIE-40GB, driver 580.173.02. Cases A-F are the reproduction matrix from #225, 3 repetitions each.

- Full path (torch 2.8.0+cu128), v2.7.1 code base:

| build | A `0` | B `1` | C `1,0` | D `1,0` + set_device(1) | E `0,1` + set_device(1) | F `0,1` |
|---|---|---|---|---|---|---|
| v2.7.1 official binary | ok 3/3 | host pid error 3/3 | host pid error 3/3 | host pid error 3/3 | ok 3/3 | ok 3/3 |
| v2.7.1 built from source (control) | ok 3/3 | host pid error 3/3 | host pid error 3/3 | host pid error 3/3 | ok 3/3 | ok 3/3 |
| v2.7.1 + this fix | ok 3/3 | ok 3/3 | ok 3/3 | ok 3/3 | ok 3/3 | ok 3/3 |

- Driver-API probe (ctypes `cuInit`, no CUDA runtime involved), `main`:

| build | A | B | C | D | E | F |
|---|---|---|---|---|---|---|
| main unpatched | ok 3/3 | host pid error 3/3 | host pid error 3/3 | host pid error 3/3 | ok 3/3 | ok 3/3 |
| main + this fix | ok 3/3 | ok 3/3 | ok 3/3 | ok 3/3 | ok 3/3 | ok 3/3 |

- Case C verbose with the fix: `current processes num = 1 2` -> `hostPid=581730` (v2.7.1 build) / `hostPid=1150676` (main build)
- Leak probe (holder process with `CVD=1,0`, 15 s): no lingering probe context during or after the run
- A clean clone of this branch builds in `nvidia/cuda:13.3.0-cudnn-devel-ubi8` (the `make build-in-docker` image)
- Why two validation paths: on our nodes, libvgpu built from current `main` (patched or not) fails CUDA runtime entry-point negotiation for every torch build we tried (cudart 12.6 / 12.8 / 13.0, all ending in `Found no NVIDIA driver`), while the official v2.7.1 binary works and direct driver-API calls also work. That regression is unrelated to this change (it reproduces on unpatched `main`) and we will report it separately. Because of it, the full-workload validation uses a v2.7.1 backport of this patch, and `main` is validated through the driver-API probe path. The probe path was cross-checked against the official v2.7.1 binary, where it reproduces the same B/C/D failures as the torch matrix.

### CODE REVIEW

**Comment by mesutoezdil** [comment]
This is being closed because it does not comply with the contribution guidelines.
